In [1]:
from postagger.utils.common import timeit, get_data_path
from postagger.utils.preprocess import load_save_preprocessed_data
from postagger.utils.decoder import CompData
import collections

In [2]:
# features occurrences
train_path = get_data_path('train.wtag')
train_sentences = CompData(train_path)
preprocessor = load_save_preprocessed_data('train.pickle', train_sentences)

In [3]:
preprocessor.pdict.keys()

dict_keys(['nextword-f107', 'number_inside', 'trigram-f103', 'previousword-f106', 'unigram-f105', 'bigram-f104', 'wordtag-f100', 'prefix-f102', 'capital_inside', 'starting_capital', 'suffix-f101'])

# exploration

In [4]:
def count(tuple_list, show_top=50):
    counter_obj = collections.Counter(tuple_list)
    total = sum(counter_obj.values())
    total_uni = len(counter_obj.items())
    most_common = counter_obj.most_common(50)
    return total, total_uni, most_common

In [5]:
def explore_all(pdict):
    for feature_key, tuples in pdict.items():
        print(feature_key)
        total, total_uni, most_common = count(tuples)
        print("Total: " + str(total))
        print("Unique: " + str(total_uni))
        for elem in most_common:
            print(elem)
        print()
        print()

In [6]:
explore_all(preprocessor.pdict)

nextword-f107
Total: 116815
Unique: 30777
(('the', 'IN'), 2793)
(('.', 'NN'), 1861)
((',', 'NN'), 1847)
(('of', 'NN'), 1731)
((',', 'NNP'), 1642)
(('.', 'NNS'), 1020)
((',', 'NNS'), 924)
(('a', 'IN'), 806)
(("'s", 'NNP'), 751)
(('to', 'NN'), 658)
(('in', 'NN'), 652)
(('.', 'NNP'), 639)
(('%', 'CD'), 624)
(('and', 'NN'), 598)
(('of', 'NNS'), 521)
(('million', 'CD'), 515)
(('the', ','), 450)
((',', 'CD'), 429)
(('the', 'VB'), 420)
(('and', 'NNS'), 419)
(('and', 'NNP'), 397)
(('for', 'NN'), 385)
(('is', 'NN'), 381)
((',', 'RB'), 369)
(('in', 'NNS'), 362)
(("'s", 'NN'), 341)
(("''", ','), 339)
(('the', 'VBD'), 326)
(('.', 'CD'), 324)
(('to', 'NNS'), 311)
(('$', 'IN'), 299)
(("''", '.'), 279)
(('to', 'VBN'), 278)
(('be', 'MD'), 278)
(('billion', 'CD'), 272)
(('by', 'VBN'), 269)
(('that', 'NN'), 264)
(('said', 'NNP'), 261)
(('are', 'NNS'), 248)
(('a', ','), 244)
(('and', ','), 243)
(('company', 'DT'), 242)
(('which', ','), 239)
(('to', 'JJ'), 232)
(('a', 'VB'), 232)
(('the', 'TO'), 226)
(('a

# Summary

Word / tag: unique 15415, keep top 50 (or greater than 228)

Prefix: unique 21525, most common are vastly NN/IN/DT, (pick top 20?)

Suffix: unique 12311, most common are NN/VB, (pick top 20?)

Unigram: 44 unique then keep all

Bigram: 1023 unique, keep 10% (top 100)

Trigram: 7810 unique, keep 10% (top 780)

Previous word: 32095 unique, keep 1% (top 320)

Next word: 30777 unique, keep 1% (top 307)

number_inside: 6 unique, keep top 3

capital inside: 21 unique, keep top 3

starting_capital: 33 unqiue, keep top 10

## Prefix:

In [25]:
counter_obj = collections.Counter(preprocessor.pdict['prefix-f102']).most_common(500)

In [32]:
[x for x in counter_obj if x[0][1] not in ['NN', 'NNS', 'NNPS', 'NNP', 'VBD', 'IN', 'DT', 'CD', '-LRB-', '-RRB-', 'WDT']],

([(('s', 'JJ'), 810),
  (('w', 'MD'), 755),
  (('c', 'JJ'), 625),
  (('f', 'JJ'), 616),
  (('a', 'RB'), 506),
  (('p', 'JJ'), 506),
  (('l', 'JJ'), 472),
  (('m', 'JJ'), 455),
  (('wil', 'MD'), 454),
  (('wi', 'MD'), 454),
  (('will', 'MD'), 454),
  (('s', 'VBZ'), 362),
  (('o', 'JJ'), 347),
  (('b', 'VBN'), 344),
  (('co', 'JJ'), 343),
  (('th', 'PRP'), 311),
  (('a', 'JJ'), 311),
  (('t', 'PRP'), 311),
  (('the', 'PRP'), 311),
  (('wo', 'MD'), 301),
  (('woul', 'MD'), 301),
  (('wou', 'MD'), 301),
  (('e', 'JJ'), 290),
  (('s', 'VB'), 279),
  (('r', 'JJ'), 277),
  (('say', 'VBZ'), 269),
  (('sa', 'VBZ'), 269),
  (('says', 'VBZ'), 269),
  (('i', 'JJ'), 268),
  (('h', 'VBP'), 266),
  (('al', 'RB'), 264),
  (('c', 'VB'), 261),
  (('ha', 'VBP'), 259),
  (('fi', 'JJ'), 258),
  (('hav', 'VBP'), 257),
  (('have', 'VBP'), 257),
  (('t', 'JJ'), 255),
  (('c', 'VBN'), 249),
  (('r', 'VB'), 248),
  (('n', 'JJ'), 245),
  (('be', 'VBN'), 242),
  (('s', 'VBN'), 235),
  (('they', 'PRP'), 228),
  ((

Alot of junk in that feature, even after filtering, possibly add its tuples manually (e.g., 'un', 'a', 'im')

# suffix

In [35]:
counter_obj = collections.Counter(preprocessor.pdict['suffix-f101']).most_common(100)

In [36]:
[x for x in counter_obj]

[(('s', 'NNS'), 7250),
 (('e', 'NN'), 3208),
 (('t', 'NN'), 2652),
 (('d', 'VBD'), 2235),
 (('es', 'NNS'), 2203),
 (('ng', 'VBG'), 2016),
 (('ing', 'VBG'), 2016),
 (('g', 'VBG'), 2016),
 (('d', 'VBN'), 2004),
 (('ed', 'VBN'), 1912),
 (('n', 'NN'), 1684),
 (('y', 'NN'), 1655),
 (('r', 'NN'), 1654),
 (('ed', 'VBD'), 1404),
 (('s', 'VBZ'), 1340),
 (('rs', 'NNS'), 1336),
 (('n', 'NNP'), 1321),
 (('e', 'JJ'), 1241),
 (('on', 'NN'), 1238),
 (('l', 'JJ'), 1212),
 (('ts', 'NNS'), 1209),
 (('y', 'RB'), 1194),
 (('ion', 'NN'), 1163),
 (('e', 'VB'), 1158),
 (('e', 'NNP'), 1112),
 (('ly', 'RB'), 1091),
 (('al', 'JJ'), 1055),
 (('t', 'IN'), 1037),
 (('er', 'NN'), 1031),
 (('t', 'JJ'), 1021),
 (('nt', 'NN'), 969),
 (('tion', 'NN'), 948),
 (('r', 'NNP'), 918),
 (('g', 'NN'), 889),
 (('s', 'NNP'), 859),
 (('ers', 'NNS'), 857),
 (('ng', 'NN'), 856),
 (('.', 'NNP'), 851),
 (('ing', 'NN'), 843),
 (('n', 'CD'), 830),
 (('y', 'NNP'), 826),
 (('lion', 'CD'), 812),
 (('on', 'CD'), 812),
 (('ion', 'CD'), 812)

Top 100 is enough.